File: CSharp/AoC2025/Days/Day10.cs
using Microsoft.Z3;

namespace AoC2025.Days;

public static class Day10
{
    public static long PartOne()
    {
        List<(int[] lights, int[][] buttons, int[] joltage)> data = ParseData();
        long total = 0;

        foreach ((int[] lights, int[][] buttons, int[] _) in data)
        {
            int n = buttons.Length;
            int min = int.MaxValue;
            int max = (int)Math.Pow(2, n);
            for (int i = 1; i < max; i++)
            {
                var combo = Enumerable.Range(0, n)
                    .Where(n => ((i >> n) & 1) != 0)
                    .Select(n => buttons[n]);
                if (MatchLights(combo, lights))
                    min = Math.Min(min, combo.Count());
            }

            total += min;
        }

        return total;
    }

    public static long PartTwo()
    {
        List<(int[] lights, int[][] buttons, int[] joltage)> data = ParseData();
        
        long total = 0;

        foreach ((int[] _, int[][] buttons, int[] joltage) in data)
        {
            total += SolveWithZ3(joltage, buttons);
        }

        return total;
    }

    // Helper methods

    private static bool MatchLights(IEnumerable<int[]> combo, int[] lights)
    {
        int length = lights.Length;
        int[] pattern = new int[length];

        foreach (int[] button in combo)
            foreach (int b in button)
                pattern[b] = pattern[b] == 0 ? 1 : 0; // Toggle switch

        for (int i = 0; i < length; i++)
            if (pattern[i] != lights[i])
                return false;

        return true;
    }

    // Z3 implementation

    public static long SolveWithZ3(int[] target, int[][] buttons)
    {
        using var ctx = new Context();
        Optimize opt = ctx.MkOptimize();

        int numButtons = buttons.Length;
        int numLights = target.Length;

        IntExpr[] x = new IntExpr[numButtons];
        for (int i = 0; i < numButtons; i++)
            x[i] = ctx.MkIntConst($"x_{i}");

        for (int light = 0; light < numLights; light++)
        {
            ArithExpr sum = ctx.MkInt(0);
            for (int btn = 0; btn < numButtons; btn++)
                if (buttons[btn].Any(b => b == light))
                    sum = ctx.MkAdd(sum, x[btn]);

            opt.Add(ctx.MkEq(sum, ctx.MkInt(target[light])));
        }

        for (int i = 0; i < numButtons; i++)
            opt.Add(ctx.MkGe(x[i], ctx.MkInt(0)));

        ArithExpr total = ctx.MkAdd(x);
        opt.MkMinimize(total);

        if (opt.Check() != Status.SATISFIABLE)
            throw new Exception("No solution found");

        Model model = opt.Model;

        long sumPresses = 0;
        for (int i = 0; i < numButtons; i++)
        {
            IntNum val = model.Evaluate(x[i], true) as IntNum;
            sumPresses += val.Int64;
        }

        return sumPresses;
    }

    
    // Data methods

    private static IEnumerable<string[]> GetData()
        => File.ReadAllLines("../../Data/2025/day10.txt")
            .Select(line => line.Split(' ').ToArray());

    private static List<(int[], int[][], int[])> ParseData()
    {
        var data = GetData();
        List<(int[], int[][], int[])> parsed = [];

        foreach (var line in data)
        {
            int[] lights = ParseLights(line[0]);
            int[][] buttons = ParseButtons(line[1..^1]);
            int[] joltage = ParseJoltage(line[^1]);

            parsed.Add((lights, buttons, joltage));
        }

        return parsed;
    }

    private static int[] ParseLights(string lights)
    {
        int[] parsed = new int[lights.Length - 2]; //Ignore brackets.

        for (int i = 0; i < lights.Length - 2; i++)
            parsed[i] = lights[i + 1] == '#' ? 1 : 0;

        return parsed;
    }

    private static int[][] ParseButtons(string[] buttons) => buttons
            .Select(b => 
                b[1..^1]
                    .Split(',')
                    .Select(s => int.Parse(s))
                    .ToArray()
            )
            .ToArray();

    private static int[] ParseJoltage(string joltage)  
        => joltage[1..^1]
            .Split(',')
            .Select(s => int.Parse(s))
            .ToArray();
}

File: JavaScript/AoC2025/days/day10.js
import fs from "fs";
import path from "path";
import { init } from "z3-solver";

const { Context } = await init();
const Z3 = Context();
const { Int, Optimize } = Z3;

export default class Day10 {
    constructor() {
        const filePath = path.resolve("../../Data/2025/day10.txt");
        const data = fs.readFileSync(filePath, "utf-8");
        this.machines = data
            .split("\n")
            .map(line => line.trim())
            .filter(line => line.length > 0)
            .map(line => this.parseLine(line));
    }

    parseLine(line) {
        const parts = line.split(" ");
        const lights = parts[0].slice(1, -1).split("").map(c => c === "#" ? 1 : 0);
        const buttons = parts.slice(1, -1).map(b => b.slice(1, -1).split(",").map(Number));
        const joltage = parts[parts.length - 1].slice(1, -1).split(",").map(Number);
        return [lights, buttons, joltage];
    }

    part1() {
        let total = 0;
        for (const [lights, buttons] of this.machines) {
            const n = buttons.length;
            let min = Infinity;
            for (let i = 1; i < (1 << n); i++) {
                const combo = [];
                for (let j = 0; j < n; j++) if ((i >> j) & 1) combo.push(buttons[j]);
                if (this.matchLights(combo, lights)) min = Math.min(min, combo.length);
            }
            total += min;
        }
        return total;
    }

    matchLights(combo, lights) {
        const pattern = Array(lights.length).fill(0);
        combo.forEach(btn => btn.forEach(b => pattern[b] = 1 - pattern[b]));
        return pattern.every((val, i) => val === lights[i]);
    }

    async part2() {
        let total = 0;
        for (const [_, buttons, joltage] of this.machines) {
            total += await this.solveWithZ3(joltage, buttons);
        }
        return total;
    }

    async solveWithZ3(target, buttons) {
        const numButtons = buttons.length;
        const numLights = target.length;

        const x = Array.from({ length: numButtons }, (_, i) => Int.const(`x_${i}`));

        const opt = new Optimize();

        for (let light = 0; light < numLights; light++) {
            const affecting = x.filter((_, j) => buttons[j].includes(light));
            const sum = affecting.reduce((acc, xi) => acc.add(xi), Int.val(0));
            opt.add(sum.eq(Int.val(target[light])));
        }

        for (let xi of x) opt.add(xi.ge(Int.val(0)));

        let totalExpr = x[0];
        for (let i = 1; i < x.length; i++) totalExpr = totalExpr.add(x[i]);
        opt.minimize(totalExpr);

        const status = await opt.check();
        if (status !== "sat") throw new Error("No solution found");

        const model = opt.model();
        let sumPresses = 0n;
        for (let xi of x) {
            const val = model.eval(xi).toString();
            sumPresses += BigInt(val);
        }

        return Number(sumPresses);
    }

    async printResults() {
        console.log("Day 10:");
        console.log(`Part 1: ${this.part1()}`);
        console.log(`Part 2: ${await this.part2()}`);
    }
}
